src/checkers/inference/model/ComparisonConstraint.java

```diff
@@ -0,0 +1,140 @@
+                default:
+                	throw new BugInCF("There are no defined ComparableOperationKind "
+                            + "for the given com.sun.source.tree.Tree.Kind: " + kind);
+            }
+        }
+
+        public String getSymbol() {
+            return opSymbol;
+        }
+    }
+
+    private ComparisonConstraint(ComparisonOperationKind operation, Slot left, Slot right,
+            ComparisonVariableSlot result, AnnotationLocation location) {
+        super(Arrays.asList(left, right, result), location);
+        this.left = left;
+        this.right = right;
+        this.operation = operation;
+        this.result = result;
+    }
+
+    private ComparisonConstraint(ComparisonOperationKind operation, Slot left, Slot right, 
+    		ComparisonVariableSlot result) {
+        super(Arrays.asList(left, right, result));
+        this.left = left;
+        this.right = right;
+        this.operation = operation;
+        this.result = result;
+    }
+    
+    protected static Constraint create(ComparisonOperationKind operation, Slot left, Slot right,
+    		ComparisonVariableSlot result, AnnotationLocation location, QualifierHierarchy realQualHierarchy) {
+        if (operation == null || left == null || right == null) {
+            throw new BugInCF("Create comparable constraint with null argument. "
+                    + "Operation: " + operation + " Subtype: "
+                    + left + " Supertype: " + right);
+        }
+        if (location == null || location.getKind() == AnnotationLocation.Kind.MISSING) {
+            throw new BugInCF(
+                    "Cannot create an ComparisonConstraint with a missing annotation location.");
+        }
+
+        return new ComparisonConstraint(operation, left, right, result, location);
+    }
+
+    @Override
+    public <S, T> T serialize(Serializer<S, T> serializer) {
+        return serializer.serialize(this);
+    }
+
+    public ComparisonOperationKind getOperation() {
+        return operation;
+    }
+
+    public Slot getLeft() {
+        return left;
+    }
+
+    public Slot getRight() {
+        return right;
+    }
+    
+    public ComparisonVariableSlot getResult() {
+        return result;
+    }
+
+    @Override
+    public int hashCode() {
+        int code = 1;
+        code = code + ((left == null) ? 0 : left.hashCode());
+        code = code + ((right == null) ? 0 : right.hashCode());
+        code = code + ((operation == null) ? 0 : operation.hashCode());
+        return code;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj)
+            return true;
+        if (obj == null)
+            return false;
+        if (getClass() != obj.getClass())
+            return false;
+        ComparisonConstraint other = (ComparisonConstraint) obj;
+        if (left.equals(other.left) && right.equals(other.right) 
+        		&& operation.equals(other.operation)) {
+            return true;
+        } else {
```

Comment:
Should we care about cases like `a < b` and `b > a`?

When is `equals` called in inference?

If we insert an entry into HashMap or HashSet, is it the hashcode that is compared to determine the membership?